lua/starfall/sflib.lua

```diff
@@ -701,48 +675,230 @@ if SERVER then
 		return sf
 	end
 
+	local function getCodeCache(ply, filename, hash, codeSize)
+		local cacheEntry = codecache[ply][filename] or {}
+		for k,v in pairs(cacheEntry) do
+			if v.hash == hash and v.size == codeSize then
+				return v
+			end
+		end
+		return nil
+	end
+
+	local function addCodeCache(ply, filename, hash, codeSize)
+		if codecache[ply][filename] == nil then
+			codecache[ply][filename] = {}
+		end
+		local cacheEntry = codecache[ply][filename]
+		local cache = { filename = filename, hash = hash, size = codeSize, time = CurTime() }
+		table.insert(cacheEntry, cache)
+		return cache
+	end
+
+	local function sweepCodeCache()
+		for ply, cachedFiles in pairs(codecache) do
+			for fname, entries in pairs(cachedFiles) do
+				for k, entry in pairs(entries) do
+					if CurTime() - entry.time >= (10 * 60) then
+						codecache[ply][fname][k] = nil
+						--print("Removing expired cache entry: " .. fname .. ", hash: " .. entry.hash)
+					end
+				end
+			end
+		end
+	end
+
+	net.Receive( "starfall_upload", function ( len, ply )
+		local updata = uploaddata[ ply ]
+		if not updata then
+			ErrorNoHalt( "SF: Player " .. ply:GetName() .. " tried to upload code without being requested (expect this message multiple times)\n" )
+			return
+		end
+
+		local readType = net.ReadUInt(2)
+		if readType == 0 then
+			-- Start of file.
+			local fname = net.ReadString()
+			local hash = net.ReadString()
+			local codeSize = net.ReadUInt(18)
+			--print("Receiving file: " .. fname .. " (Hash: " .. hash .. ")")
+			updata.files[fname] = ""
+			addCodeCache(ply, fname, hash, codeSize)
+		elseif readType == 1 then
+			-- Chunk
+			local fname = net.ReadString()
+			local data = net.ReadString()
+			updata.files[fname] = updata.files[fname] .. data
+			--print("Received chunk for: " .. fname .. ", size: " .. #data .. " bytes")
+		elseif readType == 2 then
+			-- End of file
+			local fname = net.ReadString()
+			local hash = net.ReadString()
+			local code = updata.files[fname]
+			local cache = getCodeCache(ply, fname, hash, #code)
+			cache.code = code
+		elseif readType == 3 then
+			-- End of list
+			updata.callback( updata.mainfile, updata.files )
+			uploaddata[ ply ] = nil
+		else
+			error("Unexpected read type: " .. tostring(readType))
+		end
+
+		-- Also cleanup cached entries older than 10 minutes, if they
+		-- will be used again the time updates.
+		if readType == 3 then
+			sweepCodeCache()
+		end
+
+	end )
+	
+	net.Receive( "starfall_uploadlist", function(len, ply)
+
+		local updata = uploaddata[ ply ]
+		if not updata then
+			ErrorNoHalt( "SF: Player " .. ply:GetName() .. " tried to upload code without being requested (expect this message multiple times)\n" )
+			return
+		end
+
+		local missing = {}
+		local mainfile = net.ReadString()
+		local count = net.ReadUInt(16)
```

Comment:
count might want replacing with a more descriptive word. Like num_of_files, seeing as the net interactions aren't exactly documented..